test/e2e/canary_test.go

```diff
@@ -127,6 +127,64 @@ func TestCanaryRoute(t *testing.T) {
 	}
 }
 
+// TestCanaryRouteClearsSpecHost verifies that the operator clears the canary
+// route's spec.host field (which requires deleting and recreating the route) if
+// spec.host gets set.
+//
+// This is a serial test because it modifies the canary route.
+func TestCanaryRouteClearsSpecHost(t *testing.T) {
+	t.Log("Waiting for the default IngressController to be available...")
+	if err := waitForIngressControllerCondition(t, kclient, 5*time.Minute, defaultName, defaultAvailableConditions...); err != nil {
+		t.Fatal(err)
+	}
+
+	var canaryRoute routev1.Route
+	canaryRouteName := controller.CanaryRouteName()
+	t.Log("Getting the canary route...")
+	if err := wait.PollImmediate(1*time.Second, 1*time.Minute, func() (bool, error) {
+		if err := kclient.Get(context.TODO(), canaryRouteName, &canaryRoute); err != nil {
+			t.Log(err)
+			return false, nil
+		}
+
+		return true, nil
+	}); err != nil {
+		t.Fatal(err)
+	}
+
+	if v := canaryRoute.Spec.Host; len(v) != 0 {
+		t.Fatalf("Expected canary route to have empty spec.host, found %q.", v)
+	}
+
+	const bogusHost = "foo.bar"
+	canaryRoute.Spec.Host = bogusHost
+	t.Log("Setting a bogus spec.host...")
+	if err := kclient.Update(context.TODO(), &canaryRoute); err != nil {
+		t.Fatal(err)
+	}
```

Comment:
I think we need to specifically call the function `updateCanaryRoute` to test this.

Reply:
This is an end-to-end test, not a unit test.  All it needs to do is the following:

1. Get the canary route.
2. Update its `spec.host` field.
3. Verify that the operator clears the field.

The fact that clearing the field requires deleting and recreating the route is irrelevant for this test.  If the operator succeeds in clearing `spec.host`, it's good.  Is there some case that this test fails to cover?